subsys/net/lib/sockets/sockets_select.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2018 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <sys/time.h>
+
+#include <net/socket.h>
+
+/* Get size, in elements, of an array within a struct. */
+#define STRUCT_MEMBER_ARRAY_SIZE(type, field) ARRAY_SIZE(((type *)0)->field)
+
+/* Outputs results in predefined word_idx and bit_mask vars. */
+#define FD_SET_CALC_OFFSETS(set) { \
+	unsigned int b_idx = fd % (sizeof(set->bitset[0]) * 8); \
+	word_idx = fd / (sizeof(set->bitset[0]) * 8); \
+	bit_mask = 1 << b_idx; \
+	}
+
+void ZSOCK_FD_ZERO(zsock_fd_set *set)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(set->bitset); i++) {
+		set->bitset[i] = 0;
+	}
+}
+
+int ZSOCK_FD_ISSET(int fd, zsock_fd_set *set)
+{
+	u32_t word_idx, bit_mask;
+
+	if (fd < 0 || fd >= ZSOCK_FD_SETSIZE) {
+		return 0;
+	}
+
+	FD_SET_CALC_OFFSETS(set);
+
+	return (set->bitset[word_idx] & bit_mask) != 0;
+}
+
+void ZSOCK_FD_CLR(int fd, zsock_fd_set *set)
+{
+	u32_t word_idx, bit_mask;
+
+	if (fd < 0 || fd >= ZSOCK_FD_SETSIZE) {
+		return;
+	}
+
+	FD_SET_CALC_OFFSETS(set);
+
+	set->bitset[word_idx] &= ~bit_mask;
+}
+
+void ZSOCK_FD_SET(int fd, zsock_fd_set *set)
+{
+	u32_t word_idx, bit_mask;
+
+	if (fd < 0 || fd >= ZSOCK_FD_SETSIZE) {
+		return;
+	}
+
+	FD_SET_CALC_OFFSETS(set);
+
+	set->bitset[word_idx] |= bit_mask;
+}
+
+int zsock_select(int nfds, zsock_fd_set *readfds, zsock_fd_set *writefds,
+		 zsock_fd_set *exceptfds, struct timeval *timeout)
+{
+	struct pollfd pfds[CONFIG_NET_SOCKETS_POLL_MAX];
+	int i, res, poll_timeout;
+	int num_pfds = 0;
+	int num_selects = 0;
+	int fd_no = 0;
+
+	for (i = 0; i < STRUCT_MEMBER_ARRAY_SIZE(zsock_fd_set, bitset); i++) {
```

Comment:
Why do you need this STRUCT_MEMBER_ARRAY_SIZE() macro at all, just use ARRAY_SIZE()?
You are not using the new macro for example in ZSOCK_FD_ZERO().

Reply:
> Why do you need this STRUCT_MEMBER_ARRAY_SIZE() macro at all, just use ARRAY_SIZE()?
The general idea was discussed earlier with @rlubos:

ARRAY_SIZE() takes an array variable as a parameter. 

STRUCT_MEMBER_ARRAY_SIZE() takes a *type* name and *field* name as parameters.

Reply:
> Why do you need this STRUCT_MEMBER_ARRAY_SIZE() macro at all, just use ARRAY_SIZE()?

So again, to elaborate to the essence of it. If ARRAY_SIZE was used, it would look like `ARRAY_SIZE(readfds->bitset)`. But readfds can be NULL. "Whoa, you dereference a NULL pointer!", thinks a quickly-skimming human or limited-wit bot like Coverity. To avoid alerts like that, I introduced that STRUCT_MEMBER_ARRAY_SIZE() macro. Yes, underneath it still deals with a NULL pointer (again, similarly how `offsetof()` is implemented). But that's the whole idea - we abstract "unsafe" operation behind a macro, so the uses of that macro don't cause alerts.

I would be ready to replace it with ARRAY_SIZE() now, but I'm sure that sooner or later we'll hit a tool which will false-positive it, and will need to deal with that. So, I'd prefer it to stay as "dealt with" already.